src/poly/dcrt/matrix/dcrt_poly_matrix.rs

```diff
@@ -111,68 +114,35 @@ impl PolyMatrix for DCRTPolyMatrix {
-        let bit_length = self.params.modulus_bits();
-        let new_nrow = self.nrow * bit_length;
-        let new_matrix = Self::new_empty(&self.params, new_nrow, self.ncol);
-        let (row_offsets, col_offsets) = block_offsets(0..self.nrow, 0..self.ncol);
-        parallel_iter!(row_offsets.iter().tuple_windows().collect_vec()).for_each(
-            |(cur_block_row_idx, next_block_row_idx)| {
-                parallel_iter!(col_offsets.iter().tuple_windows().collect_vec()).for_each(
-                    |(cur_block_col_idx, next_block_col_idx)| {
-                        let self_block_polys = self.block_entries(
-                            *cur_block_row_idx..*next_block_row_idx,
-                            *cur_block_col_idx..*next_block_col_idx,
-                        );
-                        let block_row_len = next_block_row_idx - cur_block_row_idx;
-                        let block_col_len = next_block_col_idx - cur_block_col_idx;
-                        let new_entries: Vec<Vec<DCRTPoly>> = (0..block_row_len)
-                            .flat_map(|i| {
-                                let decompositions: Vec<Vec<DCRTPoly>> = (0..block_col_len)
-                                    .map(|j| {
-                                        let poly = &self_block_polys[i][j];
-                                        poly.decompose(&self.params)
-                                    })
-                                    .collect();
-                                (0..bit_length)
-                                    .map(move |k| {
-                                        decompositions
-                                            .iter()
-                                            .map(|decomposed| decomposed[k].clone())
-                                            .collect::<Vec<_>>()
-                                    })
-                                    .collect::<Vec<Vec<DCRTPoly>>>()
-                            })
-                            .collect();
-                        // This is secure because the modified entries are not overlapped among
-                        // threads
-                        unsafe {
-                            new_matrix.replace_block_entries(
-                                cur_block_row_idx * bit_length..next_block_row_idx * bit_length,
-                                *cur_block_col_idx..*next_block_col_idx,
-                                new_entries,
-                            );
-                        }
-                    },
-                );
-            },
-        );
+        let gadget_vector = Self::gen_dcrt_gadget_vector(params);
+        debug_assert_eq!(gadget_vector.col_size(), params.modulus_digits());
+        gadget_vector.concat_diag(&vec![&gadget_vector; size - 1])
+    }
+
+    fn decompose(&self, base_bits: Option<u32>) -> Self {
+        let base_bits = base_bits.unwrap_or(self.params.base_bits());
+        let digits_len =
```

Comment:
Call this variable as `log_base_q` to be coherent with the rest of the codebase?